src/webview/message/handler.ts

```diff
@@ -173,27 +182,50 @@ export class MessageToWebviewHandlerImpl implements MessageToWebviewHandler {
     return extractNumberFromString(lineNumberElement.textContent);
   }
 
-  private hideViewedFiles(diffContainer: HTMLElement, viewedState: ViewedState) {
+  private async hideViewedFiles(diffContainer: HTMLElement, viewedState: ViewedState) {
     const viewedToggles = diffContainer.querySelectorAll<HTMLInputElement>(
       Diff2HtmlCssClassElements.Input__ViewedToggle
     );
     for (const toggle of Array.from(viewedToggles)) {
       const fileName = this.getDiffElementFileName(toggle);
-      if (fileName && viewedState[fileName]) toggle.click();
+      if (fileName && viewedState[fileName]) {
+        const diffHash = await this.getDiffHash(toggle);
+        if (diffHash === viewedState[fileName]) {
+          toggle.click();
+        } else {
+          toggle.classList.add(CHANGED_SINCE_VIEWED);
+        }
+      }
     }
   }
 
-  private sendFileViewedMessage(toggleElement: HTMLInputElement): void {
+  private async getDiffHash(diffElement: HTMLElement) {
+    const fileContainer = this.getDiffFileContainer(diffElement);
+    const fileContent = fileContainer?.querySelector(Diff2HtmlCssClassElements.Div__DiffFileContent)?.innerHTML;
+    if (!fileContent) return null;
+
+    const fileContentUtf8 = encoder.encode(fileContent);
+    const hash = await window.crypto.subtle.digest("SHA-1", fileContentUtf8);
+    const hashHex = Array.from(new Uint8Array(hash))
+      .map((x) => x.toString(16).padStart(2, "0"))
+      .join("");
+
+    return hashHex;
```

Comment:
The hashing code uses browser/webview API (`window.crypto.subtle`) so I'm not sure it fits in `shared/extract.ts`, I've put it in `webview/message/hash.ts`, let me know if there's a better place for it. Outside of `extract.ts` it doesn't need to be named `extract...` because it doesn't actually do any extracting, so I'd propose `getSha1Hash()`.

Unit testing would be great, do you have any webview tests already? The function won't work in the existing jest test framework because the crypto API (or indeed the global `window` object) is not available there.

Reply:
`message/hash.ts` sounds great!

How about mocking `window.crypto`? After all, we can assume the way it works.

Reply:
I found node's `crypto.webcrypto` which implements the needed API; test added. It puts `webcrypto` into `window.crypto`, please have a look at that in `hash.test.ts`.

An alternative to using `any` and disabling type checks would be to add the following before the assignment:
```ts
declare namespace global {
  let window: {
    crypto: typeof webcrypto;
  };
}
```

But we don't otherwise use `global.window` in the test file so the stricter typing brings no value.

Reply:
LGTM, thanks 😃